controllers/progressiverollout_controller.go

```diff
@@ -61,5 +69,93 @@ func (r *ProgressiveRolloutReconciler) Reconcile(ctx context.Context, req ctrl.R
 func (r *ProgressiveRolloutReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&deploymentskyscannernetv1alpha1.ProgressiveRollout{}).
+		Watches(
+			&source.Kind{Type: &argov1alpha1.Application{}},
+			handler.EnqueueRequestsFromMapFunc(r.requestsForApplicationChange)).
+		Watches(
+			&source.Kind{Type: &corev1.Secret{}}, handler.EnqueueRequestsFromMapFunc(r.requestsForSecretChange)).
 		Complete(r)
 }
+
+func (r *ProgressiveRolloutReconciler) requestsForApplicationChange(o client.Object) []reconcile.Request {
+
+	/*
+		We trigger a reconciliation loop on an Application event if:
+		- the Application owner is referenced by a ProgressiveRollout object
+	*/
+
+	var requests []reconcile.Request
+	var list deploymentskyscannernetv1alpha1.ProgressiveRolloutList
+	ctx := context.Background()
+
+	app, ok := o.(*argov1alpha1.Application)
+	if !ok {
+		err := fmt.Errorf("expected application, got %T", o)
+		r.Log.Error(err, "failed to convert object to application")
+		return nil
+	}
+
+	if err := r.List(ctx, &list); err != nil {
+		r.Log.Error(err, "failed to list ProgressiveRollout")
+		return nil
+	}
+
+	for _, pr := range list.Items {
+		if pr.HasOwnerReference(app.GetOwnerReferences()) {
+			requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{
+				Namespace: pr.Namespace,
+				Name:      pr.Name,
+			}})
+		}
+	}
+
+	return requests
+}
+
+func (r *ProgressiveRolloutReconciler) requestsForSecretChange(o client.Object) []reconcile.Request {
+
+	/*
+		We trigger a reconciliation loop on a Secret event if:
+		- the Secret is an ArgoCD cluster, AND
+		- there is an Application targeting that secret/cluster, AND
+		- that Application owner is referenced by a ProgressiveRollout object
+	*/
+
+	var requests []reconcile.Request
+	var prList deploymentskyscannernetv1alpha1.ProgressiveRolloutList
+	var appList argov1alpha1.ApplicationList
+	ctx := context.Background()
+
+	s, ok := o.(*corev1.Secret)
+	if !ok {
+		err := fmt.Errorf("expected secret, got %T", o)
+		r.Log.Error(err, "failed to convert object to secret")
+		return nil
+	}
+
+	if !internal.IsArgoCDCluster(s.GetLabels()) {
+		return nil
+	}
+
+	if err := r.List(ctx, &prList); err != nil {
+		r.Log.Error(err, "failed to list ProgressiveRollout")
+		return nil
+	}
+	if err := r.List(ctx, &appList); err != nil {
+		r.Log.Error(err, "failed to list Application")
+		return nil
+	}
+
+	for _, pr := range prList.Items {
+		for _, app := range appList.Items {
+			if app.Spec.Destination.Server == string(s.Data["server"]) && pr.HasOwnerReference(app.GetOwnerReferences()) {
```

Comment:
As this is a loop. If I have 2 apps watched by 1 pr, or 2 apps watched by 2 pr, will the output on this function be longer than 1?

i.e. should we deduplicate this list?

Reply:
You can't have 2 Apps watched by 2 PRs, because this mean we have 2 PRs watching the same AppSet and then which PR "wins" in deciding how to update the Applications? We probably need a validation hook for this

Regarding the 2 Apps watched by 1 PR that happens all the time, because an AppSet can generate multiple applications.

Is this more clear? I think I don't get the issue you are pointing at.